Add PC_RotateQuaternion

elemoine opened the pull request:
This PR adds `PC_RotateQuaternion` to rotate patches based on rotation quaternions.

Commit:
Add PC_RotateQuaternion(p pcpatch)

This adds a PC_RotateQuarternion(p pcpatch) function that can be used to rotate
a patch based on a unit quaternion.

Many contributions from @mbredif.

elemoine commented:
This is WIP. DO NOT MERGE.

mbredif commented on lib/pc_patch.c:
```diff
@@ -659,3 +659,70 @@ pc_patch_transform(const PCPATCH *patch, const PCSCHEMA *new_schema)
+        PCPOINT *opt = pc_pointlist_get_point(opl, i);
+        PCPOINT *npt = pc_point_make(schema);
+
+        pc_point_get_double_by_name(opt, dimx, &(*vec)[0]);
```
for performance, you should do the dimx, dimy, dimz lookups only once per patch

elemoine replied:
You mean do a lookup for the very first point and then use indices?

mbredif replied:
Yes, I mean `dim = pc_schema_get_dimension_by_name(schema, name);` once then do `pc_point_get_double(pt, dim, d);` for each point.

mbredif commented on lib/pc_patch.c:
```diff
@@ -659,3 +659,70 @@ pc_patch_transform(const PCPATCH *patch, const PCSCHEMA *new_schema)
+
+        pc_pointlist_add_point(npl, npt);
+
+        pcfree((void *)rotatedvec);
```
should be freed after the loop

elemoine replied:
Right, will change.

mbredif commented on lib/pc_patch.c:
```diff
@@ -659,3 +659,70 @@ pc_patch_transform(const PCPATCH *patch, const PCSCHEMA *new_schema)
+
+    schema = patch->schema;
+
+    opl = pc_pointlist_from_patch(patch);
```
can't we modify patch in place ? here, values of other dimensions are left unassigned

elemoine replied:
`PC_RotateQuaternion` returns a new `PcPatch` so I don't think an in-place modification is possible. Regarding, the other dimensions, I think they should be copied in the new `PcPatch`.

mbredif replied:
I think the pointlist is a temporary copy already so that it can be modified in place, but that should be checked obviously.

elemoine replied:
Yep, you may be right. I'll check.

mbredif commented on lib/pc_patch.c:
```diff
@@ -659,3 +659,70 @@ pc_patch_transform(const PCPATCH *patch, const PCSCHEMA *new_schema)
+    PCPATCH_UNCOMPRESSED *paout;
+    const PCSCHEMA *schema;
+    PCPOINTLIST *opl, *npl;
+    const PCMAT44 *qmat;
```
for a linear rotation, a PCMAT33 and a VEC3 could be used instead

elemoine replied:
Right, will change this.

mbredif commented on lib/pc_matrix.c:
```diff
@@ -0,0 +1,80 @@
+* Create the 4x4 matrix associated to the (qw, qx, qy, qz) quaternion. Assume
+* that the quaternion is a unit quaternion.
+*/
+PCMAT44 *
```
why do you use a PCMAT44 for a normalized quaternion ? I suggest to use a linear PCMAT33 for a normalized quaternion and a projective PCMAT44 for an unnormalized quaternion, replacing 1's by |q|^2 on the diagonal.

elemoine replied:
As already said, I'll use `PCMAT33` and `PCVEC3` only for this first patch.

elemoine commented on pgsql/pointcloud.sql.in:
```diff
@@ -318,6 +318,10 @@ CREATE OR REPLACE FUNCTION PC_PatchTransform(p pcpatch, pcid int4)
     LANGUAGE 'c' IMMUTABLE STRICT;
 
+CREATE OR REPLACE FUNCTION PC_RotateQuaternion(p pcpatch, qw float8, qx float8, qy float8, qz float8, dimx text, dimy text, dimz text)
+    RETURNS pcpatchAS 'MODULE_PATHNAME', 'pcpatch_rotate_quaternion'
```
Missing space before `AS` here.

elemoine commented on lib/pc_patch.c:
```diff
@@ -659,3 +659,70 @@ pc_patch_transform(const PCPATCH *patch, const PCSCHEMA *new_schema)
+        pc_point_get_double_by_name(opt, dimx, &(*vec)[0]);
+        pc_point_get_double_by_name(opt, dimy, &(*vec)[1]);
+        pc_point_get_double_by_name(opt, dimz, &(*vec)[2]);
+        (*vec)[3] = 0;
```
To be removed when a 3x3 matrix will be used.

elemoine commented:
@mbredif I addressed your (good) comments. I need to write more tests.

elemoine commented:
@mbredif feel free to review this again.

Commit:
Add PC_RotateQuarternion(p pcpoint)

This adds a PC_RotateQuaternion(p pcpoint) function that can be used to rotate
a point based on a unit quaternion.

Many contributions from @mbredif.

mbredif commented:
Looks good !
3 comments, addressing the first one being sufficient for now in my opinion : 
- A comment is needed for future maintainers of the patch version of PC_RotateQuaternion, explaining that points iterated from pointlists extracted from uncompressed patches are just pointers to the raw patch buffer, so that updating the points directly changes the patch payload.
- x and y dimensions should ideally default to the `x_position` and `y_position`. `z_position` is however not provided. I propose to defer that to another issue (https://github.com/LI3DS/pointcloud/issues/32)
- point rotation through quaternions can be optimized in terms of floating point operations and memory allocation (see : [Quaternions_and_spatial_rotation](https://en.wikipedia.org/w/index.php?title=Quaternions_and_spatial_rotation&section=26#Comparison_with_other_representations_of_rotations) / Used methods 2)

elemoine commented:
> A comment is needed for future maintainers of the patch version of PC_RotateQuaternion, explaining that points iterated from pointlists extracted from uncompressed patches are just pointers to the raw patch buffer, so that updating the points directly changes the patch payload.

I added a comment.

> x and y dimensions should ideally default to the x_position and y_position. z_position is however not provided. I propose to defer that to another issue

Agree.

elemoine commented:
> point rotation through quaternions can be optimized in terms of floating point operations and memory allocation (see : Quaternions_and_spatial_rotation / Used methods 2)

Please be more specific. I don't know undertand what method you're referring to.

elemoine commented:
I think I found it. I'll give it a try.

mbredif commented:
The most efficient way to rotate a vector with a quaternion is ![{\vec {v}}_{\text{new}}={\vec {v}}+2{\vec {r}}\times ({\vec {r}}\times {\vec {v}}+w{\vec {v}})](https://wikimedia.org/api/rest_v1/media/math/render/svg/3cd7aa3064e12b9321bc210e1e72c7ef488416e7) ([wikipedia](https://en.wikipedia.org/w/index.php?title=Quaternions_and_spatial_rotation&section=26#Performance_comparisons)), which needs 15 multiplications and 15 additions to evaluate (or 18 muls and 12 adds if the factor of 2 is done via multiplication.), compared to the optimal 21 multiplications ( = 12 + 9 ) and 18 add/subtraction ( = 12 + 6 ) with the method of this PR (matrix creation + matrix application).

Considering that the proposed matrix creation method `pc_matrix_create_from_quaternion` is currently not optimized (6 pow, 30 multiplications, 12 add/sub rather than the optimal 0 pow,12 mult,12 add/sub ), I suggest to accept this PR as is and open an issue for this optimization if that becomes a bottleneck, ok?

elemoine commented:
I am currently trying to make Method 2 work, but it does not – my unit tests fail. I don't understand why.

elemoine commented:
@mbredif please take a look at https://github.com/elemoine/pointcloud/commit/0b6b2e558688d945d481fceb7bfbdac1a31f869d if you have some time. I cannot find what it is not correct and why it doesn't work.

elemoine commented:
http://gamedev.stackexchange.com/questions/28395/rotating-vector3-by-a-quaternion is also relevant.

mbredif commented:
Thinking back, actually, you construct the matrix once and apply it for each point so its construction cost is amortized and your solution is indeed optimal as soon as you have more than a few points.  The direct quaternion-based rotation ( elemoine@0b6b2e5 ) will thus only be relevant for version where the quaternion changes for each point... Sorry for the confusion.

elemoine commented:
> Thinking back, actually, you construct the matrix once and apply it for each point so its construction cost is amortized and your solution is indeed optimal as soon as you have more than a few points. The direct quaternion-based rotation ( elemoine@0b6b2e5 ) will thus only be relevant for version where the quaternion changes for each point... Sorry for the confusion.

Good point. That being said I still don't understand why https://github.com/elemoine/pointcloud/commit/0b6b2e5 doesn't work.

elemoine commented:
Before we merge this I'd like to do another change. I think the `pc_patch|point_rotate_quaternion` functions could place `vec` and `rvec` on the stack instead of using `malloc`.

elemoine commented:
> Before we merge this I'd like to do another change. I think the pc_patch|point_rotate_quaternion functions could place vec and rvec on the stack instead of using malloc.

Done.

mbredif commented:
Could you review mbredif/pointcloud/rotate that enforce const correctness (point rotation is performed in place, which is kind of weird compared to patch rotation which is returning a new patch) and computes rotation matrix on the stack.

elemoine commented:
@mbredif https://github.com/mbredif/pointcloud/commit/60cd5c94cd14cc08420e399d7587adb4ba5361c9 in your branch includes many unrelated things.

mbredif commented:
My mistake, I cleaned that up in mbredif@a92c265.

elemoine commented:
@mbredif I updated the patch based on your commits. This looks good. Thanks for your great suggestions.

mbredif reviewed (changes requested):
Looks good, just a little concern in the comment above.

elemoine commented:
> Looks good, just a little concern in the comment above.

What comment?

mbredif commented on pgsql/pc_editor.c:
```diff
@@ -0,0 +1,111 @@
+    serpatch = pc_patch_serialize(patch_out, NULL);
+
+    pc_patch_free(patch_in);
+    pc_patch_free(patch_out);
```
only required if patch_in has been cloned or decompressed rather than modified in place

elemoine replied:
As I just commented the output patch is always a new patch. So the user of `pc_patch_rotate_quaternion` is responsible for freeing that output patch.

mbredif commented on lib/pc_patch.c:
```diff
@@ -660,3 +660,66 @@ pc_patch_set_schema(const PCPATCH *patch, const PCSCHEMA *new_schema)
+    pc_matrix_set_from_quaternion(qmat, qw, qx, qy, qz);
+
+    if ( patch->type == PC_NONE )
+    {
```
My understanding is that you need to pc_patch_uncompressed_clone the patch to ensure that you can modify it.
- going back and forth to a pointlist does not make that logic apparent
- do you really need to clone if the patch is not readonly ? (is the patch readonly in this PR's use case ?)

elemoine replied:
In the case where the input patch is not compressed we could do the rotation in place. But for a consistent API, where the returned patch is alway a new patch, I decided to clone the input patch and do the rotation in that cloned patch.

mbredif replied:
Concerning the api, that pattern is already used in the lib, for instance :
- [Conditionnal clone](https://github.com/pgpointcloud/pointcloud/blob/3929653f51296f5dd5fe5997c8c4c5d46419cb50/pgsql/pc_access.c#L579)
- [Conditionnal free](https://github.com/pgpointcloud/pointcloud/blob/3929653f51296f5dd5fe5997c8c4c5d46419cb50/pgsql/pc_access.c#L622)

a `pc_patch_uncompressed_clone` function (and its siblings) would enable more readable code and optimized execution.

elemoine replied:
Agree with `pc_patch_uncompressed_clone`, but I did not want to put too much in that patch.

elemoine replied:
> Concerning the api, that pattern is already used in the lib, for instance

So do you mean that my implementation in fine w.r.t. this?

elemoine commented:
@mbredif more comments on this PR? Should we merge it?

mbredif commented:
Thanks @elemoine for this PR.
I deferred my little concerns to #38, #39.
